android/titanium/src/java/org/appcelerator/titanium/TiBaseActivity.java

```diff
@@ -760,6 +762,38 @@ public void onChanged(TiActivitySafeAreaMonitor monitor)
 					windowProxy.fireSafeAreaChangedEvent();
 				}
 			}
+
+			@Override
+			public void onKeyboardChanged(boolean isVisible, int width, int height, Insets keyboardSize)
+			{
+				if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.Q) {
+					if (isVisible != keyboardVisible) {
+						TiWindowProxy windowProxy = TiBaseActivity.this.window;
+						if (windowProxy != null) {
+							windowProxy.fireSafeAreaChangedEvent();
+						}
+						KrollDict kdAll = new KrollDict();
+						KrollDict kdFrame = new KrollDict();
+
+						KrollDict kdX = new KrollDict();
+						kdX.put("left", keyboardSize.left);
+						kdX.put("right", keyboardSize.right);
+
+						KrollDict kdY = new KrollDict();
+						kdX.put("top", keyboardSize.top);
+						kdX.put("bottom", keyboardSize.bottom);
+						kdFrame.put("x", kdX);
+						kdFrame.put("y", kdY);
+						kdFrame.put("height", keyboardSize.bottom);
+						kdFrame.put("width", width - keyboardSize.left - keyboardSize.right);
+						kdAll.put("keyboardFrame", kdFrame);
+						kdAll.put("animationDuration", 0);
+						
+						tiApp.fireAppEvent("keyboardframechanged", kdAll);
```

Comment:
Last one: We should guard if there is actually an event listener, so the whole logic is not called for cases where the client does not even listen to the event.

Reply:
sure, I'll add a check.
Always wondered: does it make a difference if there is no listener and you call it?

Reply:
The fix will apply for all `Ti.App` event listeners

Reply:
looks like it didn't like that last change. Some tests failed so I'm testing it without the hasListener

Reply:
Can't you just add the check here?

Reply:
I'm currently re-arranging to logic since it will crash apps at the moment. The check will be in place after that

Reply:
edit: I have to go into TiApplication and do the check there since it's on `Ti.App` and not the `Window`. Guess I have to add some more checks so it won't crash :disappointed:

Reply:
:facepalm: 
https://github.com/tidev/titanium_mobile/blob/a16d76c767025a67a7375edd9da70a4d99edb597/android/titanium/src/java/org/appcelerator/kroll/KrollProxy.java#L781
it does the check already :rofl: